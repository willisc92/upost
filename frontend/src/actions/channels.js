import API from "../utils/API";

// SET_CHANNELS
export const setChannels = (channels) => ({
    type: "SET_CHANNELS",
    channels
});

// START_SET_CHANNELS
export const startSetChannels = () => {
    return (dispatch) => {
        API.get("channels/", {
            params: {
                user: localStorage.getItem("user_id")
            }
        }).then((result) => {
            dispatch(setChannels(result.data));
        });
    };
};

export const addChannel = (channel) => {
    return (dispatch) => {
        API.post("channels/", {
            ...channel,
            user: localStorage.getItem("user_id")
        })
            .then((result) => {
                dispatch(startSetChannels());
            })
            .catch((err) => {
                alert(err);
            });
    };
};
