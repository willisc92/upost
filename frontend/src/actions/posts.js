import API, { setContentToForm, resetContentType } from "../utils/API";
import { getCurrentUser } from "./auth";
import moment from "moment";

/**
 * POST_START.
 * action generator.
 * Resets the error and sets loading status.
 *
 * @returns {Object} object to reset error and loading status
 */
export const postStart = () => ({
    type: "POST_START"
});

/**
 * SET_POSTS.
 * action generator.
 * adds user's posts to store.
 *
 * @param {Object[]} posts posts to add to store
 * @returns {Object} object to add posts to store
 */
export const setPosts = (posts) => ({
    type: "SET_POSTS",
    posts
});

/**
 * START_SET_POSTS.
 * from API gets the current user's posts and adds to store.
 *
 * @returns dispatch function
 */
export const startSetMyPosts = () => {
    return (dispatch) => {
        dispatch(postStart());
        getCurrentUser()
            .then((res) => {
                API.get("posts/", {
                    params: {
                        user: res.data.id
                    }
                })
                    .then((result) => {
                        dispatch(postSuccess());
                        dispatch(setPosts(result.data));
                    })
                    .catch((err) => {
                        dispatch(postFail(err));
                    });
            })
            .catch((err) => {
                console.log(JSON.stringify(err));
            });
    };
};

/**
 * SET_INTEREST_RANDOM_POSTS.
 * action generator.
 * resets the interestRandomPosts in store to be an empty array.
 *
 * @returns {Object} object to reset interestRandomPosts
 */
export const setInterestRandomPosts = () => {
    return { type: "SET_INTEREST_RANDOM_POSTS" };
};

/**
 * ADD_INTEREST_RANDOM_POSTS.
 * action generator.
 * adds a new interest and associated random posts to store.
 *
 * @param {Object} interestPosts contains tag and an array of posts
 * @returns {Object} object to add interest and posts to interestRandomPosts
 */
export const addInterestRandomPosts = (interestPosts) => {
    return { type: "ADD_INTEREST_RANDOM_POSTS", interestPosts };
};

/**
 * START_SET_INTEREST_RANDOM_POSTS.
 * resets the interestRandomPosts store.
 * for each interest provided gets random posts from API and adds to store.
 *
 * @param {Object[]} interests an array of interest objects or interest tags
 * @returns {Function} dispatch function
 */
export const startSetInterestRandomPosts = (interests) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            dispatch(setInterestRandomPosts());
            // if we have a list of interest objects
            if (interests.length > 0 && !!interests[0].interest_tag) {
                for (let i = 0; i < interests.length; i++) {
                    API.get("random-posts/", {
                        params: {
                            interest: interests[i].interest_tag
                        }
                    }).then((result) => {
                        if (result.data.length > 0) {
                            dispatch(addInterestRandomPosts({ tag: interests[i].interest_tag, posts: result.data }));
                        }
                    });
                }
            }
            // if we have a list of interest tags
            else {
                for (let i = 0; i < interests.length; i++) {
                    API.get("random-posts/", {
                        params: {
                            interest: interests[i]
                        }
                    }).then((result) => {
                        if (result.data.length > 0) {
                            dispatch(addInterestRandomPosts({ tag: interests[i], posts: result.data }));
                        }
                    });
                }
            }
            dispatch(postSuccess());
            resolve(true);
        });
    };
};

/**
 * START_GET_POST.
 * from API gets a specifc post and adds to store.
 *
 * @param {number} id post id to get
 * @returns {Promise}
 */
export const startGetPost = (id) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            API.get("posts/", {
                params: {
                    post_id: id
                }
            })
                .then((result) => {
                    dispatch(postSuccess());
                    dispatch(setPosts(result.data));
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * GET_POST.
 * from API gets a specific post, but does not add to store.  Does not dispatch any actions.
 *
 * @param {number} id post id to get
 */
export const getPost = (id) => {
    return new Promise((resolve, reject) => {
        API.get("posts/", {
            params: {
                post_id: id
            }
        })
            .then((result) => {
                resolve(result.data);
            })
            .catch((err) => {
                reject(err);
            });
    });
};

/**
 * ADD_POST.
 * sends post request to API to add a new user created post.
 * upon sucessful post reloads user posts.
 *
 * @param {Object} post post object to add to API
 * @returns {Promise}
 */
export const addPost = (post) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            if (post.toString() == "[object FormData]") {
                setContentToForm();
            }
            dispatch(postStart());
            API.post(`posts/`, post)
                .then((result) => {
                    if (post.toString() == "[object FormData]") {
                        resetContentType();
                    }
                    dispatch(postSuccess());
                    resolve(result);
                })
                .catch((err) => {
                    resetContentType();
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * EDIT_POST.
 * sends put request to API to edit a user's post.
 * upon sucessful post reloads user posts.
 *
 * @param {number} id post id to edit
 * @param {Object} updates updates to be made to the post
 */
export const editPost = (id, updates) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            if (updates.toString() == "[object FormData]") {
                setContentToForm();
            }
            dispatch(postStart());
            API.patch(`posts/${id}/`, updates)
                .then((result) => {
                    if (updates.toString() == "[object FormData]") {
                        resetContentType();
                    }
                    dispatch(postSuccess());
                    resolve(result);
                })
                .catch((err) => {
                    resetContentType();
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * POST_FAIL.
 * action generator.
 * adds error to the store.
 *
 * @param {Object} error the error
 * @returns {Object} object to set error
 */
export const postFail = (error) => ({
    type: "POST_FAIL",
    error
});

/**
 * POST_SUCCESS.
 * action generator.
 * resets loading and error status.
 *
 * @returns {Object} object to reset loading and error status
 */
export const postSuccess = () => ({
    type: "POST_SUCCESS"
});

/**
 * CLEAR_POSTS.
 * action generator.
 * clears posts in store
 *
 * @returns {Object} object to reset posts
 */
export const clearPosts = () => ({
    type: "CLEAR_POSTS"
});

/**
 * searchPosts
 * action dispatcher - searches for posts
 * @param {string} text - text to search posts by
 * @returns {Promise} to be handled
 */
export const searchPosts = (text) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            API.get(`posts/?search=${text}`)
                .then((result) => {
                    dispatch(postSuccess());
                    dispatch(setPosts(result.data));
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * deletePost
 * action dispatcher - soft deletes a post
 * @param {number} id - id of post to delete
 * @param {Object} post - Post object to delete
 * @returns {Promise} to be handled
 */
export const deletePost = (id) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            const updates = {
                deleted_flag: true,
                deletion_date: moment().toDate()
            };
            API.patch(`posts/${id}/`, updates)
                .then((result) => {
                    dispatch(postSuccess());
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * restorePost
 * action dispatcher - restores a post
 * @param {number} id - id of post to restore
 * @param {Object} post - Post object to restore
 * @returns {Promise} to be handled
 */
export const restorePost = (id) => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            const updates = {
                deleted_flag: false,
                deletion_date: null
            };
            API.patch(`posts/${id}/`, updates)
                .then((result) => {
                    dispatch(postSuccess());
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * getNonInterestPosts
 * action dispatcher - gets posts that are NOT relavent to your interests
 * @returns {Promise} to be handled
 */
export const getNonInterestPosts = () => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            API.get(`non-interest-posts/`)
                .then((result) => {
                    dispatch(postSuccess());
                    dispatch(setNonInterestPosts(result.data));
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));

                    reject(err);
                });
        });
    };
};

/**
 * setNonInterestPosts
 * action generator - sets posts not relavent to interests
 * @returns {Object} Action object
 */
export const setNonInterestPosts = (nonInterestPosts) => ({
    type: "SET_NON_INTEREST_POSTS",
    nonInterestPosts
});

/**
 * getRandomNonInterestPost
 * action dispatcher - Gets a random post that is not relavent to interests
 * @returns {Promise} Promise to handle
 */
export const getRandomNonInterestPost = () => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            API.get(`random-post/`)
                .then((result) => {
                    dispatch(postSuccess());
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};

/**
 * getCommunityPosts
 * action dispatcher - Gets posts only according to interests
 * @returns {Promise} Promise to handle
 */
export const getCommunityPosts = () => {
    return (dispatch) => {
        return new Promise((resolve, reject) => {
            dispatch(postStart());
            API.get(`community-posts/`)
                .then((result) => {
                    dispatch(postSuccess());
                    dispatch(setPosts(result.data));
                    resolve(result);
                })
                .catch((err) => {
                    dispatch(postFail(err));
                    reject(err);
                });
        });
    };
};
