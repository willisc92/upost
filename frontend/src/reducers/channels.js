// Channels Reducers
const channelsDefaultState = [];

export default (state = channelsDefaultState, action) => {
    switch (action.type) {
        case "SET_CHANNELS":
            return {
                channels: action.channels
            };
        default:
            return state;
    }
};
