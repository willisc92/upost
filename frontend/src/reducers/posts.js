// Post Reducers
export const postsDefaultState = { error: null, loading: false, posts: [], interestRandomPosts: [] };

export default (state = postsDefaultState, action) => {
    switch (action.type) {
        case "POST_START":
            return {
                ...state,
                error: null,
                loading: true
            };
        case "SET_POSTS":
            return {
                ...state,
                posts: action.posts
            };
        case "POST_FAIL":
            return {
                ...state,
                error: action.error,
                loading: false
            };
        case "POST_SUCCESS":
            return {
                ...state,
                error: null,
                loading: false
            };
        case "SET_INTEREST_RANDOM_POSTS":
            return {
                ...state,
                interestRandomPosts: []
            };
        case "ADD_INTEREST_RANDOM_POSTS":
            return {
                ...state,
                interestRandomPosts: [...state.interestRandomPosts, action.interestPosts]
            };
        default:
            return state;
    }
};
