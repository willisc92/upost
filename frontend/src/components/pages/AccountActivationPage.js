import React from "react";
import { startActivation } from "../../actions/activate";

class AccountActivationPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            inProgress: true,
            isValid: undefined
        };
    }

    componentDidMount() {
        const uid = this.props.match.params.uid;
        const token = this.props.match.params.token;
        startActivation(uid, token)
            .then((result) => {
                this.setState(() => {
                    return { inProgress: false, isValid: true };
                });
            })
            .catch((error) => {
                console.log("error", JSON.stringify(error, null, 2));
                this.setState(() => {
                    return { inProgress: false, isValid: false };
                });
            });
    }

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        {this.state.inProgress ? (
                            <h1 className="page-header__title">
                                Your <span>UPost</span> account is currently being activated!
                            </h1>
                        ) : this.state.isValid ? (
                            <h1 className="page-header__title">
                                Thank you for your email confirmation. Please login and complete your registration.
                            </h1>
                        ) : (
                            <h1 className="page-header__title">The activation link is invalid</h1>
                        )}
                    </div>
                </div>
            </div>
        );
    }
}

export default AccountActivationPage;
