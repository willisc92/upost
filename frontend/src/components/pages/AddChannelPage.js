import React from "react";
import ChannelForm from "../forms/ChannelForm";
import { connect } from "react-redux";
import { addChannel } from "../../actions/channels";

class AddChannelPage extends React.Component {
    constructor(props) {
        super(props);
    }

    onSubmit = (channel) => {
        this.props
            .addChannel({ ...channel })
            .then(() => this.props.history.push("/myChannels"))
            .catch(() => {});
    };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add a Channel</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ChannelForm onSubmit={this.onSubmit} />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addChannel: (channel) => dispatch(addChannel(channel))
});

export default connect(
    null,
    mapDispatchToProps
)(AddChannelPage);
