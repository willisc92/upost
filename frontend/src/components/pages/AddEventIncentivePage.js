import React from "react";
import { addIncentivePackage } from "../../actions/incentivePackage";
import { getCurrentUser } from "../../actions/auth";
import IncentiveForm from "../forms/IncentiveForm";
import { connect } from "react-redux";
import { startSetEvent, clearEvents } from "../../actions/events";

import Box from "@material-ui/core/Box";
import Container from "@material-ui/core/Container";
import Typography from "@material-ui/core/Typography";
import Button from "@material-ui/core/Button";

class AddEventIncentivePage extends React.Component {
    constructor(props) {
        super(props);
    }

    componentWillMount() {
        this.props.clearEvents();
        const event_id = this.props.match.params.id;
        getCurrentUser()
            .then((res) => {
                this.props
                    .startSetEvent(event_id)
                    .then((event_res) => {
                        if (res.data.username !== event_res.data.event_owner) {
                            this.props.history.push(`/myChannels`);
                        }
                    })
                    .catch((err) => {
                        console.log(JSON.stringify(err, null, 2));
                    });
            })
            .catch((err) => {
                console.log(JSON.stringify(err, null, 2));
            });
    }

    onSubmit = (incentive) => {
        const post_id = this.props.event.post;
        this.props
            .addIncentivePackage(incentive)
            .then((res) => this.props.history.push(`/post-events/${post_id}`))
            .catch((err) => {
                console.log(err);
            });
    };

    goBack = () => {
        const event_id = this.props.match.params.id;
        const post_id = this.props.event.post;
        this.props.history.push(`/myPosts/${post_id}/events/${event_id}/edit`);
    };

    goToIncentive = () => {
        const event_id = this.props.match.params.id;
        this.props.history.push(`/myEvents/${event_id}/editIncentive`);
    };

    render() {
        const read_only = !!this.props.event && this.props.event.deleted_flag;
        const existing_incentive = !!this.props.event && !!this.props.event.event_incentive;

        return (
            !!this.props.event && (
                <div>
                    <Box bgcolor="secondary.main" py={3}>
                        <Container fixed>
                            <Typography gutterBottom variant="h2">
                                Add an Incentive Package to Event:{" "}
                                <Typography variant="inherit" display="inline" color="primary">
                                    {this.props.event && this.props.event.event_title}
                                </Typography>
                            </Typography>

                            {existing_incentive ? (
                                <Box paddingBottom={2}>
                                    <Typography variant="h2" color="error" gutterBottom>
                                        There is already an existing incentive tied to this event.
                                    </Typography>
                                    <Button variant="contained" color="primary" onClick={this.goToIncentive}>
                                        Go to Incentive
                                    </Button>
                                </Box>
                            ) : (
                                read_only && (
                                    <Typography variant="h2" color="error" gutterBottom>
                                        The event this incentive will be tied to is deleted. Restore it before adding an
                                        incentive.
                                    </Typography>
                                )
                            )}
                            <Box>
                                <Button variant="contained" color="primary" onClick={this.goBack}>
                                    Go to Event
                                </Button>
                            </Box>
                        </Container>
                    </Box>
                    <Container fixed>
                        <IncentiveForm
                            onSubmit={this.onSubmit}
                            nextStep="Save"
                            event={this.props.event}
                            fromEvent={true}
                            read_only={read_only || existing_incentive}
                        />
                    </Container>
                </div>
            )
        );
    }
}

const mapStateToProps = (state) => ({
    event: state.events.events.length !== 0 ? state.events.events : null
});

const mapDispatchToProps = (dispatch) => ({
    clearEvents: () => dispatch(clearEvents()),
    addIncentivePackage: (incentive) => dispatch(addIncentivePackage(incentive)),
    startSetEvent: (id) => dispatch(startSetEvent(id))
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(AddEventIncentivePage);
