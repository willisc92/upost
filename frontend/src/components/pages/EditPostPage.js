import React from "react";
import { startGetPost, editPost, clearPosts, deletePost, restorePost } from "../../actions/posts";
import { connect } from "react-redux";
import PostForm from "../forms/PostForm";
import { getCurrentUser } from "../../actions/auth";
import Typography from "@material-ui/core/Typography";
import Box from "@material-ui/core/Box";
import Container from "@material-ui/core/Container";
import Button from "@material-ui/core/Button";
import CustomStepper from "../CustomStepper";

export class EditPostPage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            steps: [],
            activeStep: undefined
        };
    }

    componentDidMount() {
        this.props.clearPosts();
        const post_id = this.props.match.params.id;
        getCurrentUser()
            .then((res) => {
                this.props
                    .startGetPost(post_id)
                    .then((post_res) => {
                        if (res.data.username !== post_res.data[0].user) {
                            this.props.history.push("/myChannels");
                        } else {
                            this.setState(() => ({
                                steps: [
                                    { label: "Bulletin Boards", onClick: this.moveToBulletinBoards },
                                    {
                                        label: `Bulletin Board`,
                                        onClick: this.goToChannel
                                    },
                                    { label: `Post: ${this.props.post.post_title}`, onClick: this.moveToPost },
                                    { label: "Edit Post", onClick: null },
                                    { label: "?", onClick: null }
                                ],
                                activeStep: 3
                            }));
                        }
                    })
                    .catch((err) => {
                        console.log(JSON.stringify(err, null, 2));
                    });
            })
            .catch((err) => {
                console.log(JSON.stringify(err, null, 2));
            });
    }

    onSubmit = (post) => {
        const post_id = this.props.match.params.id;
        this.props
            .editPost(post_id, post)
            .then((result) => {
                if (post.toString() == "[object FormData]") {
                    this.props.history.push(`/channels/${post.get("channel")}`);
                } else {
                    this.props.history.push(`/channels/${post.channel}`);
                }
            })
            .catch((err) => {
                console.log(JSON.stringify(err, null, 2));
            });
    };

    onEditIncentiveClick = () => {
        this.props.history.push(`/myPosts/${this.props.match.params.id}/editIncentive`);
    };

    onAddIncentiveClick = () => {
        this.props.history.push(`/myPosts/${this.props.match.params.id}/addIncentive`);
    };

    onEditEventsClick = () => {
        this.props.history.push(`/post-events/${this.props.match.params.id}`);
    };

    deletePost = () => {
        const id = this.props.post.post_id;
        this.props
            .deletePost(id)
            .then(() => {
                this.goToChannel();
            })
            .catch((err) => {
                console.log(JSON.stringify(err, null, 2));
            });
    };

    restorePost = () => {
        const id = this.props.post.post_id;
        this.props
            .restorePost(id)
            .then(() => {
                this.props.history.push(`/myPosts/${id}/edit`);
            })
            .catch((err) => {
                console.log(JSON.stringify(err, null, 2));
            });
    };

    goToChannel = () => {
        const channel = this.props.post.channel;
        this.props.history.push(`/channels/${channel}`);
    };

    moveToBulletinBoards = () => {
        this.props.history.push("/myChannels");
    };

    moveToPost = () => {
        this.props.history.push(`/post/${this.props.post.post_id}`);
    };

    render() {
        const read_only_channel = !!this.props.post && this.props.post.channel_deleted_flag;
        const read_only_post = !!this.props.post && this.props.post.deleted_flag;

        return (
            !!this.props.post && (
                <div>
                    <Box bgcolor="secondary.main" py={3}>
                        <Container maxWidth="xl">
                            <Typography variant="h1" gutterBottom>
                                Edit Post:{" "}
                                <Typography variant="inherit" display="inline" color="primary">
                                    {this.props.post && this.props.post.post_title}
                                </Typography>
                            </Typography>
                            {read_only_channel ? (
                                <Box>
                                    <Typography variant="h2" color="error" gutterBottom>
                                        You must restore the bulletin board of this post before editing.
                                    </Typography>
                                    <Button color="primary" variant="contained" onClick={this.goToChannel}>
                                        Go to Bulletin Board
                                    </Button>
                                </Box>
                            ) : read_only_post ? (
                                <Box>
                                    <Typography variant="h2" color="error" gutterBottom>
                                        You must restore this post before editing.
                                    </Typography>
                                    <Button color="primary" variant="contained" onClick={this.restorePost}>
                                        Restore Post
                                    </Button>
                                </Box>
                            ) : (
                                <Box>
                                    <CustomStepper steps={this.state.steps} activeStep={this.state.activeStep} />
                                    <Button color="primary" variant="contained" onClick={this.onEditEventsClick}>
                                        Add/Edit Events
                                    </Button>{" "}
                                    {!!this.props.post.post_incentive ? (
                                        <Button color="primary" variant="contained" onClick={this.onEditIncentiveClick}>
                                            Edit Post Perk
                                        </Button>
                                    ) : (
                                        <Button color="primary" variant="contained" onClick={this.onAddIncentiveClick}>
                                            Add Perk to Post
                                        </Button>
                                    )}{" "}
                                    <Button color="primary" variant="contained" onClick={this.deletePost}>
                                        Delete Post
                                    </Button>
                                </Box>
                            )}
                        </Container>
                    </Box>
                    <Container maxWidth="xl">
                        <Box py={3}>
                            {!!this.props.post && (
                                <PostForm
                                    onSubmit={this.onSubmit}
                                    channel={this.props.post.channel}
                                    post={this.props.post}
                                    nextStep="Save Changes"
                                    read_only={read_only_channel || read_only_post}
                                />
                            )}
                        </Box>
                    </Container>
                </div>
            )
        );
    }
}

const mapStateToProps = (state) => ({
    loading: state.posts.loading,
    post: state.posts.posts[0]
});

const mapDispatchToProps = (dispatch) => ({
    clearPosts: () => dispatch(clearPosts()),
    startGetPost: (id) => dispatch(startGetPost(id)),
    editPost: (id, updates) => dispatch(editPost(id, updates)),
    deletePost: (id) => dispatch(deletePost(id)),
    restorePost: (id) => dispatch(restorePost(id))
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(EditPostPage);
