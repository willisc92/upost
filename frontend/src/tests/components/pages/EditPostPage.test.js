import React from "react";
import { shallow } from "enzyme";
import { EditPostPage } from "../../../components/pages/EditPostPage";
import channels from "../../fixtures/channels";
import posts from "../../fixtures/posts";

let wrapper, match, editPost, startGetPost, history, post;

beforeEach(() => {
    post = posts[0];
    match = {
        params: {
            id: post.post_id
        }
    };

    history = {
        push: jest.fn()
    };

    editPost = jest.fn().mockImplementation(() => {
        return Promise.resolve();
    });

    startGetPost = jest.fn().mockImplementation(() => {
        return Promise.resolve();
    });

    localStorage.setItem("user_name", post.user);

    wrapper = shallow(
        <EditPostPage match={match} post={post} editPost={editPost} startGetPost={startGetPost} history={history} />
    );
});

test("Should render page correctly", () => {
    expect(wrapper).toMatchSnapshot();
});

test("Should handle onSubmit and redirect on successful submission", async () => {
    await wrapper.instance().onSubmit(post);
    expect(editPost).toHaveBeenCalledWith(post.post_id, post);
    expect(history.push).toHaveBeenCalledWith(`/myChannels/${post.channel}`);
});
