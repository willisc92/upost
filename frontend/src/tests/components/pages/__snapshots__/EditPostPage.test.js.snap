// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Post
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <Connect(PostForm)
      channel={1}
      onSubmit={[Function]}
      post={
        Object {
          "channel": 1,
          "cost": 0,
          "deleted_flag": false,
          "email": "willisc92@hotmail.com",
          "phone_number": "4034010115",
          "post_description": "Test",
          "post_event": null,
          "post_id": 1,
          "post_timestamp": "2019-06-04T23:23:53.199984-06:00",
          "post_title": "Post 1",
          "poster_name": "Willis",
          "tags": Array [
            "Interest 1",
          ],
          "user": "willisc92",
        }
      }
    />
  </div>
</div>
`;
